tests/integration/HACKING.md

```diff
@@ -0,0 +1,468 @@
+
++ **girder_client_token**
++ **girder_api_url**
+
+These are **not** used in Girder job creation,  but instead provide each executing, consumer side, task with access to an instantiated [GirderClient](http://girder.readthedocs.io/en/latest/python-client.html#the-python-client-library). This client can be used to perform arbitrary RESTful calls against the girder instance. The URL location of the girder API is specified with ```girder_api_url``` keyword argument. The default ```girder_api_url``` will be the return value of ```girder.plugins.worker.utils.getWorkerApiUrl()```. The client will be instantiated with the ```girder_client_token``` as its authentication mechanism. This means the consumer-side client will have permissions that are consistent with the scopes applied to the token. The default ```girder_client_token``` is ```None```.
+
+An example:
+
+Producer side REST endpoint:
+```python
+from my_custom_module.tasks import some_task
+from girder.utility.model_importer import ModelImporter
+
+...
+
+@access.token
+@filtermodel(model='job', plugin='jobs')
+@describeRoute(
+    Description('Test girder client is generated and can request scoped endpoints')
+	.param('file_id', 'File Id to run "some_task" on', dataType='string')))
+def some_rest_endpoint(file_id, **params):
+	# Create a token with permissions of the current user.
+    token = ModelImporter.model('token').createToken(
+        user=self.getCurrentUser())
+
+    # Launch the asynchronous task
+	async_result = some_task.delay(file_id, girder_client_token=token)
+	
+	return async_result.job
+```
+
+Consumer side task:
+```python
+from girder_worker.app import app
+
+@app.task(bind=True)
+def some_task(self, file_id):
+	# Note: girder_client is available as an attribute of 'self'
+	# self (the task object) will be passed to the function if
+	# bind is set to True in the task decorator.
+	self.girder_client.downloadFile(file_id, '/tmp/file')
+	with open('/tmp/file', 'rb') as f:
+	    # Do something with contents of f
+
+```
+
+In this example the producer side REST endpoint creates a token with the scope of the current user, then passes the REST argument ```file_id``` and the token to the task object's ```.delay(...)``` function.  On the consumer side,  the task has access to the GirderClient instance through ```self.girder_client```. To get access to ```self``` the task must be decorated with ```@app.task(bind=True)``` (See [bound tasks](http://docs.celeryproject.org/en/latest/userguide/tasks.html#bound-tasks) for more information). 
+
+# How do I write a custom Celery task for Girder Worker?
+
+*Note that this is taken almost directly from [Writing your own plugin](http://girder-worker.readthedocs.io/en/latest/developer-docs.html#writing-your-own-plugin)*
+
+Adding additional tasks to the girder\_worker infrastructure is easy and takes three steps. 
+
+1. Creating tasks
+2. Creating a plugin class 
```

Comment:
why does it have to be a plugin?  i'd want more background on why it is a plugin and a plugin system of what?